tests/src/test/scala/akka/kafka/scaladsl/RebalanceSpec.scala

```diff
@@ -243,6 +250,212 @@ class RebalanceSpec extends SpecBase with TestcontainersKafkaLike with Inside {
       control1.isShutdown.futureValue shouldBe Done
       control2.isShutdown.futureValue shouldBe Done
     }
+
+    "no message loss during partition revocation and re-subscription" in assertAllStagesStopped {
+      // BEGIN: vals and defs
+      val topicCount = 10
+      val partitionCount = 10
+      val perPartitionMessageCount = 1000
+      val businessSleep = 10.milliseconds
+      val rebalanceEventBufferTime = 8.seconds
+      val expectedTimeToFinish = 60.seconds // with 19 seconds buffer after last message
+      // inmemory message storage along with duplicate message count
+      val messageStorage = new CMap[Int, AtomicInteger]().asScala
+      val group1 = createGroupId(1)
+      val consumerSettings = consumerDefaults
+        .withProperty(ConsumerConfig.MAX_POLL_RECORDS_CONFIG, "20") // 500 is the default value
+        //.withProperty(ConsumerConfig.PARTITION_ASSIGNMENT_STRATEGY_CONFIG, classOf[RangeAssignor].getName)
+        .withProperty(ConsumerConfig.PARTITION_ASSIGNMENT_STRATEGY_CONFIG, classOf[RoundRobinAssignor].getName)
+        .withGroupId(group1)
+      val topicMap = new CMap[String, MSet[TopicPartition]]().asScala
+      val topicIdxMap = new CMap[Int, String]().asScala
+
+      def businessFlow(clientId: String): Flow[CommittableMessage[String, String], CommittableOffset, NotUsed] = {
+        Flow.fromFunction { message =>
+          val messageVal = message.record.value.toInt
+          if (!messageStorage.contains(messageVal)) {
+            messageStorage(messageVal) = new AtomicInteger(0)
+          }
+          val duplicateCount = messageStorage(messageVal).incrementAndGet()
+          log.debug(
+            s"businessFlow offset ${message.committableOffset.partitionOffset.offset} messageId=${messageVal} topicPartition=${message.record.topic}-${message.record.partition} consumerId=${clientId} duplicateCount=${duplicateCount}"
+          )
+          // sleep to simulate expensive business logic
+          sleep(businessSleep, "business sleep time")
+          message.committableOffset
+        }
+      }
+
+      def subscribeAndConsumeMessages(clientId: String, subscription: AutoSubscription, tpCount: Int, ptCount: Int) =
+        Consumer
+          .committablePartitionedSource(consumerSettings.withClientId(clientId), subscription)
+          .mapAsyncUnordered(tpCount * ptCount) {
+            case (topicPartition, topicPartitionStream) => {
+              topicPartitionStream
+                .via(businessFlow(clientId))
+                .map(offsetOption => {
+                  log.debug(
+                    s"topicPartition ${topicPartition.topic}-${offsetOption.partitionOffset.key.partition} offset ${offsetOption.partitionOffset.offset} consumer ${clientId}"
+                  )
+                  offsetOption
+                })
+                .runWith(Committer.sink(committerDefaults))
```

Comment:
This is most likely the problem. You can't do committing in `mapAsyncUnordered` it might commit offsets in a different order than the messages are processed.
Do you get the same error if this uses `mapAsync`?

Reply:
Ah, I didn't look at it long enough. This is OK as it is applied to the stream of partitioned sources, sorry.